django/db/models/fields/__init__.py

```diff
@@ -2305,6 +2305,8 @@ class UUIDField(Field):
 
     def __init__(self, verbose_name=None, **kwargs):
         kwargs['max_length'] = 32
+        if kwargs.get("primary_key") and kwargs.get("auto_created"):
+            kwargs.setdefault("default", uuid.uuid4)
```

Comment:
This feels to me like it is attempting to fudge a non-integer auto-field (`UUIDAutoField`) which has been discussed in the past. A lot more work is needed for this as there are lots of `isinstance(…, AutoField)` checks. See the discussions and implementation in #10680 that provide the initial cleanup preparing for generic auto-fields.

Should we be allowing any field type to be used? Or should we only be allowing `AutoFieldMixin` subclasses? Should the setting be `DEFAULT_AUTOFIELD` instead of `DEFAULT_MODEL_PK`?

I feel this PR should be limited to supporting specifying an alternative to the default `AutoField`. The brief IIRC was the desire to implement a mechanism such that we can default to `BigAutoField` without causing migration headaches for existing projects. Additional auto-created field types should be handled as a separate issue.

Reply:
I like the idea of being able to specify any supported field as the primary key automatically. We support `UUIDField(primary_key=True)` right now (seemingly) without problems. I'll spend some time grepping around for the `isinstance` checks.

I'm not sure I agree that this should be handled as a different PR - you won't get an `AutoField` with `primary_key=True, auto_created=True` and without this change a `UUIDField` is pretty useless as a primary key field. To me they seem intricately linked.

Reply:
Hmm. Maybe I'm muddying the waters a little bit. Don't know.

Up till now the auto-generated field has always been an `AutoField`. In that sense it is tied to being an integer and having the value auto-generated by the database and returned via a `RETURNING` clause (or similar). I understand that you can manually override things in your model by specifying something like `UUIDField(primary_key=True)` - I have often done this too. But allowing it to take the place of `AutoField` as an auto-generated `id` field on the model is different. Hence my comments about checking for `isinstance(..., AutoField)` - there are a lot of hard-coded assumptions about the auto-generated field on a model.

I guess I'm thinking of something like `UUIDAutoField` where it is the database that actually provides the default value. We now have some support via `db_returning` for _creating_ new objects that return a value from the database. But we don't yet have the ability to specify something like `db_default` to choose a function to be executed on the database to generate that value, e.g. [`gen_random_uuid()`](https://www.postgresql.org/docs/13/functions-uuid.html) in PostgreSQL.

Maybe I'm making a mountain out of a molehill, but I just wanted to highlight this stuff for careful consideration. :male_detective: 

This change feels like a bit of a kludge so that we can get a sort-of "support auto-generated UUID primary key" :heavy_check_mark: in addition to the original concern of not using `BigAutoField` by default. (Not that I don't want it.)

Reply:
I did a quick audit of `AutoField` usages within Django, and I can find two potential issues:

1. `bulk_create` does `fields = [f for f in fields if not isinstance(f, AutoField)]`
2. `forms.models.construct_instance` does `if not f.editable or isinstance(f, models.AutoField) or f.name not in cleaned_data: continue`

I'm not sure how to continue. On one hand, flexibility is nice and I can see a use for this in specific circumstances, but you are right that there might be additional issues we cannot see.

The issue with `UUIDAutoField` and using `db_default` in general is portability. As not all databases support generating a UUID (and PG needs an extension loaded) we'd surely need to fall back to `uuid.uuid4()` in some cases. I feel like this might require a lot of foundational work.

I think we should try to make this generic (I'll add test cases for all potential issues I can see), but we have to accept that there may be bugs that are discovered through this. In general we're pretty good at not assuming much about primary key fields as we do support different kinds, so I'd be really surprised if it does.

All that being said - this does increase the scope of the proposed change by a non-trivial amount. I'm obviously more keen about AutoField -> BigAutoField than anything else, so I'm happy to drop the generalization if that's the consensus.

Reply:
> I'm not sure how to continue.

Focus on making this work for switching `AutoField` to `BigAutoField` which was the primary brief. Other than careful decisions about naming the setting to avoid having to change it later, we should be able to consider how we are going to support non-auto fields for this as part of a separate PR later. That will include determining how to address the `isinstance(…, AutoField)` refactoring that was also punted from when I fixed the auto-field inheritance.

> As not all databases support generating a UUID (and PG needs an extension loaded) …

Most do, and loading an extension wouldn't be a deal breaker. We already have to do such a thing to enable `JSONField` for SQLite. At some point we need to accept that some extra effort is needed to enable certain functionality in certain backends. Also [`gen_random_uuid()`](https://www.postgresql.org/docs/13/functions-uuid.html) will be available without an extension in PostgreSQL 13 so gradually things will get easier.

Reply:
> Focus on making this work for switching AutoField to BigAutoField which was the primary brief.

Ok, sounds good. For the implementation I am thinking that we keep the `DEFAULT_MODEL_PK` setting but raise an exception if it's not an `AutoField` subclass? I feel that it would be the most future proof approach rather than having some specific `USE_BIGINT_AUTOFIELD` setting.

I'll reduce the scope of the PR in the coming week.

Reply:
> I'll reduce the scope of the PR in the coming week.

Cool. Addressing this whole area in smaller, manageable chunks should hopefully result in ensuring we can get this landed.

> …than having some specific `USE_BIGINT_AUTOFIELD` setting.

Eww. That would be ghastly. 🤪 I can think of the following:

- `DEFAULT_MODEL_PK` - as currently implemented, but doesn't get across that it is a "field".
- `DEFAULT_AUTOFIELD` - my original suggestion, depends whether it is interpreted as an instance of `AutoField` or an "auto generated field". Perhaps `DEFAULT_AUTO_FIELD` instead…
- `DEFAULT_PK_FIELD` - could get confused with forms? Or maybe `DEFAULT_AUTO_PK_FIELD`…

Reply:
> Ok, sounds good. For the implementation I am thinking that we keep the DEFAULT_MODEL_PK setting but raise an exception if it's not an AutoField subclass? I feel that it would be the most future proof approach rather than having some specific USE_BIGINT_AUTOFIELD setting.

+1 to that.

Reply:
I've changed the PR to raise a ValueError when the chosen PK field is not a subclass of `AutoField` 👍